### PR TITLE
Update to feedback generation, implementation of demo creation and train/test seed generation, and refactoring of dataset creation

### PR DESCRIPTION
**Pull request by sabraaap**
**1) Update to feedback generation:**
a) fixed the logic for subtasks created for opening a locked door, which required not just creating a new ObjDesc instance with a key of the same colour as the door but also explicitly resetting the verifier so that the matching goal objects list is populated (this took me a while to figure out...)
b) added an action condition to the goto feedback (mostly to cover any unexpected edge cases)
c) removed the second GoTo subtask for PutNext tasks which checked if the agent went to the fixed object, as this is not required (and would potentially even incentivise suboptimal behaviour, if this is learned by the agent) to put something next to the fixed object
d) extended the messages returned when a subtask has been completed to make it implicit that 'correct' refers to a subtask, not the overall task (just in case this distinction in the resulting representations is picked up by the agent) - for the last subtask and single goto tasks (which have no subtasks), this is adjusted to refer to the task, rather than a part of the task.

**2) Implementation of demo creation**
We can now easily create demo videos of a) training episodes and b) 'made up' default episodes that efficiently illustrate the feedback mechanism, with the corresponding feedback. The latter includes two modes (using manually created action sequences) for the easiest and the hardest environment. The former can be customised to any dataset and episode in the dataset (with the option to create a dataset on the fly). The demos also include the action at each step, and the mission description.

**3) Implementation of train/test seed generation**
This allows us to generate and store a number of training seeds (by default 100 per environment level and config, which do not include the held out environment configurations used for testing) and testing seeds (by default 10 environment and config and type of OOD phenomenon - see below for more details). These are stored in a json file and can be accessed using the level and config name as keys, and in the case of the test seeds, the type of OOD phenomenon. Note that for some levels and configs, certain OOD phenomenons are not applicable - here the seed list will be empty - and that some seeds are OOD with respect to multiple OOD phenomenons. I'm not sure if we will be combining these seeds into sets or whether we'll be testing for each OOD phenomenon separately, or both? 
The phenomenons are listed below:
- _size_: seeds that involve a specific room or maze size - this is set to _rooms smaller than 5x5 and larger than 8x8_ (the default room size), and _mazes that don't conform to the default maze dimensions of _3x3_ rooms. Note that for certain configs and even levels, all seeds will be either entirely in-domain or OOD.
- _color-type_: seeds that involve a specific combination of colour and type of object as a/the goal - this combination was chosen randomly, but with the random seed used, results in the combination _yellow box_.
- _agent-loc_: seeds that involve a specific agent start position in the room and/or maze - this was again chosen at random, but with the random seed used, results in _room quadrant (1, 2) (= bottom left)_ and _maze quadrant (or room) (3, 3) (=bottom right)_.
- _object-task_: seeds that involve a specific combination of object and task - this was set to a random object for the fixed object in PutNext tasks (= the object next to which the agent needs to put another object) and with the random seed used, this results in the fixed object for PutNext tasks being _blue ball_. Note that this only applies to PutNext single task levels and configs, and sequence task levels.
- _rel-loc_: seeds that involve a specific goal location relative to the agent - this was again chosen at random, but with the random seed used, results in goal objects _on your right_. Note that this applies only to levels and configs that make use of the location descriptor for goal objects.
- _task-task_: seeds that involve specific task combinations in sequence tasks - this was set to having to _open two or more locked doors_ as (part of) the task. Note that this applies only yo levels and configs that make use of sequences and include locked doors.

**4) Refactoring dataset creation**
Due to the fact that feedback is now generated concurrently with the rest of the dataset, it seemed sensible to ensure that we can access this for every episode in the same way we can access the observations/actions/rewards for a given episode, using the dataset.episodes attribute. This required a lot of changes that would have seen me recreate 75% of the MinariDataset class in the CustomDataset subclass, so I decided to just get rid of the subclass and just implement all of the project-specific functionality directly in the MinariDataset class. I then got rid of the old CustomDataset, and instead reused this class name for the actual code for generating datasets from the generate_datasets file, moving all the execution code from generate_datasets to the make_datasets file you'd already created. This means the module and the script for making datasets now live in their respective separate files and folders. The code for Google/dopamine datasets now also lives in the MinariDataset class. I've updated all references to CustomDataset where it should now be MinariDataset (and also changed the way feedback is handled in the Collator - let me know if you think that won't work and we need to change it).

### CODE REVIEW

File: src/collator/collator.py [hunk 1]
```diff
@@ -67,13 +66,13 @@ def __init__(
 
         # store feedback as flattened array. if no feedback provided, use empty strings
         self.feedback = (
-            np.hstack(feedback)
-            if feedback is not None
+            np.hstack(custom_dataset.feedback)
+            if feedback
```
Comment (aleSuglia) [inline src/collator/collator.py:70]: Instead of using a random vector for the "empty" feedback. Why don't we generate the embedding for a fixed sentence like "no feedback available"? In this way, we avoid using a representation that might diverge too much from the original embedding space learned by sentence-bert.

File: src/collator/collator.py [hunk 2]
```diff
@@ -145,14 +146,18 @@ def _sample_batch(self, batch_size):
             s.append(
                 self._normalise_states(
                     self._pad(
```
Comment (aleSuglia) [inline src/collator/collator.py:148]: I can't see the implementation of `self._pad` but in general I would recommend using `pad_sequence` from Pytorch which is highly efficient because written in C++: https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html

File: src/dataset/seeds.py [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+            if ood_type in ["object_task", "task_task"]:
+                max_seeds_to_check = 1000
+            else:
+                max_seeds_to_check = 100
+            for seed in tqdm(range(0, 10000)):
+                if seed >= max_seeds_to_check and len(seed_list) == 0:
+                    break
+                env = gym.make(config)
+                env.reset(seed=seed)
+                if ood_type == "size":
```
Comment (aleSuglia) [inline src/dataset/seeds.py:616]: Mostly a stylistic choice: I would rewrite this long list of "if" with a single "switch". In Python: https://stackoverflow.com/questions/21962763/using-a-dictionary-as-a-switch-statement-in-python

**Review by aleSuglia — CHANGES_REQUESTED** [review]
Thanks for doing this. This looks great. Just a few comments, happy to discuss later!

File: src/collator/collator.py (hunk 1 above)
Comment (sabraaap) [reply src/collator/collator.py:70]: Good point!

File: scripts/train_agent.py [hunk 4]
```diff
@@ -24,10 +24,12 @@
 device = setup_devices(SEED, useGpu=True)
```
[the remark this answers is not in the document]
Comment (sabraaap) [reply scripts/train_agent.py:24]: On the todo list :-)

File: src/collator/collator.py (hunk 2 above)
Comment (sabraaap) [reply src/collator/collator.py:148]: @maxtaylordavies
